File: CPP/No227.cpp
/**
 * Created by Xiaozhong on 2020/8/22.
 * Copyright (c) 2020/8/22 Xiaozhong. All rights reserved.
 */

#include "iostream"
#include "string"
#include "stack"

using namespace std;

class Solution {
public:
    int calculate(string s) {
        int ans = 0, n = 0; // ans 记录最终的计算答案， n 用来将 "123" 转换为 123
        char sign = '+';    // 最初的符号，从左到右肯定是 + 开头呀
        stack<int> stack;   // 计算栈
        for (int i = 0; i < s.length(); i++) {
            if (s[i] >= '0' && s[i] <= '9') {   // 字符到数字的转换过程
                n = n * 10 + (s[i] - '0');
            }
            if (s[i] < '0' && s[i] != ' ' || i == s.length() - 1) {
                int pre;        // 再下面用来记录前一个数字
                switch (sign) {
                    case '+':   // 如果是 + ，那么就直接添加进来
                        stack.push(n);
                        break;
                    case '-':   // 如果是 - ，那么就添加一个负数
                        stack.push(-n);
                        break;
                    case '*':   // 如果是 * ，那么就将栈顶元素相乘，然后再放回来
                        pre = stack.top();
                        stack.pop();
                        stack.push(pre * n);
                        break;
                    case '/':   // 如果是 / ，那么就将栈顶元素相除，再放回来
                        pre = stack.top();
                        stack.pop();
                        stack.push(pre / n);
                        break;
                }
                sign = s[i];    // 记录符号
                n = 0;          // 清空当前转换的数字
            }
        }
        while (!stack.empty()) {    // 最后把栈里面的数字挨个加起来
            ans += stack.top();
            stack.pop();
        }
        return ans;
    }
};

int main() {
    Solution s;
    cout << s.calculate("3+2*2") << endl;
}

File: Java/src/No227.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.Deque;
import java.util.LinkedList;

public class No227 {
    public int calculate(String s) {
        // ans 用来返回计算结果，n 用来把 "123" 转换成 123
        int ans = 0, n = 0;
        Deque<Integer> stack = new LinkedList<>();
        // 记录当前计算符号状态
        char sign = '+';
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            // 遇到字符数了，把字符转成数字
            if (Character.isDigit(c)) n = n * 10 + c - '0';
            // 如果遇到有效符号 +-*/ ，那么就执行计算
            // 需要加上 i == s.length() - 1 ，来确保最后一个数字也能参与计算
            if (c < '0' && c != ' ' || i == s.length() - 1) {
                switch (sign) {
                    case '+':
                        stack.addLast(n);   // 是加号状态，就直接把数字放进来
                        break;
                    case '-':
                        stack.addLast(-n);  // 负号状态，就放进来一个负的数字
                        break;
                    case '*':
                        stack.addLast(stack.pollLast() * n);    // 先把栈顶元素取出来，把栈顶元素和当前元素相乘后，放进栈中
                        break;
                    case '/':
                        stack.addLast(stack.pollLast() / n);    // 同样把栈顶元素取出来，然后相除后放进栈中
                        break;
                }
                sign = s.charAt(i); // 更新符号状态
                n = 0;  // 清空缓存数据
            }
        }
        // 把栈中的元素挨个加起来
        while (!stack.isEmpty()) ans += stack.pollLast();
        return ans;
    }

    public static void main(String[] args) {
        String s = "3+2*2";
        System.out.println(new No227().calculate(s));
        System.out.println(new No227().calculate("1+2*5/3+6/4*2"));
    }
}
